File: sar/detector.py
import re
from typing import Optional
from sar.models import (
    RedactionCandidate, PIICategory, RedactionStatus,
    SubjectDetails, TextSpan, DetectionSettings,
)
from sar.nhs_patterns import find_regex_matches
from sar.staff_list import is_staff_name
from sar.name_detector import detect_names
from sar.pdf_parser import get_full_page_text
from sar.custom_words import get_custom_words
from sar.risk_words import check_text_for_risk

# Default confidence thresholds (used when no DetectionSettings provided)
AUTO_REDACT_THRESHOLD = 0.80
FLAG_THRESHOLD = 0.50

# Labels that indicate a name belongs to a staff member / practitioner.
# Use inline (?i:...) only for the label keywords; name-part stays case-sensitive.
_STAFF_LABEL_RE = re.compile(
    r'(?i:Practitioner|Record\s+author|Filed\s+by|Requested\s+by|'
    r'Authorised\s+by|Discontinued\s+by|Performed\s+by|Cancellation\s+reason)'
    r':\s*(?:(?:Mr|Mrs|Ms|Miss|Dr|Prof|Sister|Nurse|Rev|Mx)\.?\s+)*',
)

# Pattern to extract a staff name following a staff label.
# Stops at § (field separator), newline, "at " (org separator), or end of string.
# No re.IGNORECASE — name-part [A-Z][a-z] must remain case-sensitive.
# Handles hyphenated surnames like Smith-Jones.
_NAME_COMPONENT = r"[A-Z][a-z\']+(?:-[A-Z][a-z\']*)?"
_STAFF_NAME_EXTRACT_RE = re.compile(
    r'(?i:Practitioner|Record\s+author|Filed\s+by|Requested\s+by|'
    r'Authorised\s+by|Discontinued\s+by|Performed\s+by)\s*:\s*'
    r'((?:(?:Mr|Mrs|Ms|Miss|Dr|Prof|Sister|Nurse|Rev|Mx)\.?\s+)?'
    r'(?:' + _NAME_COMPONENT + r')(?:[ \t]+(?:' + _NAME_COMPONENT + r')){0,3})'
    r'(?=[ \t]*(?:\xa7|\n|$|at\s+[A-Z]))',
)


def normalize(text: str) -> str:
    return text.strip().lower()


def _strip_punctuation(text: str) -> str:
    """Remove trailing/leading punctuation from names."""
    return re.sub(r'^[\s\.,;:\'"-]+|[\s\.,;:\'"-]+$', '', text)


def _extract_page_staff_names(page_text: str) -> set[str]:
    """
    Extract names that appear in staff-labelled positions on this page.
    E.g. "Practitioner: Dr David Triska" → {"david triska", "dr david triska"}
    """
    staff = set()
    for m in _STAFF_NAME_EXTRACT_RE.finditer(page_text):
        raw = m.group(1).strip()
        staff.add(normalize(raw))
        # Also add without title prefix
        parts = raw.split()
        titles = {"mr", "mrs", "ms", "miss", "dr", "prof", "sister", "nurse", "rev", "mx"}
        while parts and parts[0].lower().rstrip('.') in titles:
            parts.pop(0)
        if parts:
            staff.add(normalize(" ".join(parts)))
    return staff


def _is_in_staff_label_context(page_text: str, name_start: int) -> bool:
    """
    Check if the name at name_start is directly preceded by a staff label
    (e.g. "Practitioner:", "Record author:").
    """
    # Look back up to 90 chars to allow for title prefix after the colon
    pre = page_text[max(0, name_start - 90):name_start]
    return bool(_STAFF_LABEL_RE.search(pre))


def is_subject_match(entity_text: str, subject: SubjectDetails) -> bool:
    """Check if detected text matches the data subject's known details."""
    cleaned = normalize(_strip_punctuation(entity_text))

    # Full name check
    if subject.full_name:
        if cleaned == normalize(subject.full_name):
            return True
        if normalize(subject.full_name) in cleaned:
            return True

    # First name (only match single-word entities — lone first name)
    if subject.first_name:
        entity_words = cleaned.split()
        if entity_words == [normalize(subject.first_name)]:
            return True

    # Last name: match lone surname OR full name where first word also matches the subject
    if subject.last_name:
        subj_last = normalize(subject.last_name)
        entity_words = cleaned.split()
        if subj_last in entity_words:
            if len(entity_words) == 1:
                return True  # Lone surname
            elif subject.first_name and normalize(subject.first_name) in entity_words:
                return True  # Both first & last name match subject

    # Aliases
    for alias in subject.aliases:
        if alias and cleaned == normalize(alias):
            return True

    # NHS number
    if subject.nhs_number:
        stripped_entity = re.sub(r'[\s\-]', '', entity_text)
        stripped_subject = re.sub(r'[\s\-]', '', subject.nhs_number)
        if stripped_entity == stripped_subject:
            return True

    # Date of birth
    if subject.date_of_birth and cleaned == normalize(subject.date_of_birth):
        return True

    # Phone
    if subject.phone:
        stripped_entity = re.sub(r'[\s\-\(\)]', '', entity_text)
        stripped_subject = re.sub(r'[\s\-\(\)]', '', subject.phone)
        if stripped_entity == stripped_subject:
            return True

    # Email
    if subject.email and cleaned == normalize(subject.email):
        return True

    # Address
    if subject.address:
        addr_words = set(normalize(subject.address).split())
        entity_words = set(cleaned.split())
        stopwords = {"the", "a", "an", "and", "of", "in", "at", "on", "to"}
        entity_words -= stopwords
        if entity_words and len(entity_words & addr_words) >= max(1, len(entity_words) * 0.7):
            return True

    return False


def map_text_to_spans(
    page_text: str,
    page_spans: list[TextSpan],
    char_start: int,
    char_end: int,
) -> list[TextSpan]:
    """
    Find TextSpan objects whose bounding boxes correspond to the given
    character range in page_text.
    """
    matching = []
    search_from = 0

    for span in page_spans:
        pos = page_text.find(span.text, search_from)
        if pos == -1:
            pos = page_text.find(span.text)
        if pos == -1:
            continue

        span_end = pos + len(span.text)
        search_from = pos + 1

        if pos < char_end and span_end > char_start:
            matching.append(span)

    return matching


def detect_pii(
    pdf_path: str,
    text_spans: list[TextSpan],
    subject: SubjectDetails,
    source_filename: str,
    settings: Optional[DetectionSettings] = None,
) -> list[RedactionCandidate]:
    """
    Detect third-party PII in a single PDF.
    Combines rule-based name detection with regex patterns.
    """
    if settings is None:
        settings = DetectionSettings()

    auto_thresh = settings.auto_redact_threshold
    flag_thresh = settings.flag_threshold
    enabled = set(settings.enabled_categories)

    candidates = []

    # Load custom words once for the whole document
    custom_words = get_custom_words()

    # Group spans by page
    pages: dict[int, list[TextSpan]] = {}
    for span in text_spans:
        pages.setdefault(span.page_num, []).append(span)

    for page_num, page_spans in sorted(pages.items()):
        page_text = get_full_page_text(pdf_path, page_num)
        if not page_text.strip():
            continue

        # Extract practitioner/author names from labelled fields on this page
        # (e.g. "Practitioner: Dr David Triska") for contextual staff exclusion
        page_staff_names = _extract_page_staff_names(page_text)

        # ── Name detection ────────────────────────────────────────────────
        name_matches = detect_names(page_text)

        for nm in name_matches:
            entity_text = _strip_punctuation(nm.text)
            if not entity_text:
                continue

            if is_subject_match(entity_text, subject):
                candidates.append(RedactionCandidate(
                    text=entity_text,
                    category=PIICategory.PERSON_NAME,
                    status=RedactionStatus.EXCLUDED_SUBJECT,
                    confidence=1.0,
                    page_num=page_num,
                    reason="Matched data subject",
                    source_file=source_filename,
                ))
                continue

            # Check staff list first, then contextual staff names from labelled fields
            entity_norm = normalize(entity_text)
            # Strip title from entity for contextual check
            entity_words = entity_norm.split()
            titles = {"mr", "mrs", "ms", "miss", "dr", "prof", "sister", "nurse", "rev", "mx"}
            entity_no_title = " ".join(w for i, w in enumerate(entity_words)
                                       if not (i == 0 and w.rstrip('.') in titles))

            in_staff_label = _is_in_staff_label_context(page_text, nm.start)
            in_page_staff = (entity_norm in page_staff_names or
                             entity_no_title in page_staff_names)

            if is_staff_name(entity_text) or in_staff_label or in_page_staff:
                reason = "Matched staff list"
                if in_staff_label:
                    reason = "Name in staff-labelled field (Practitioner/Author)"
                elif in_page_staff:
                    reason = "Name matched practitioner on this page"
                candidates.append(RedactionCandidate(
                    text=entity_text,
                    category=PIICategory.PERSON_NAME,
                    status=RedactionStatus.EXCLUDED_STAFF,
                    confidence=1.0,
                    page_num=page_num,
                    reason=reason,
                    source_file=source_filename,
                ))
                continue

            matched_spans = map_text_to_spans(page_text, page_spans, nm.start, nm.end)

            if matched_spans:
                x0 = min(s.x0 for s in matched_spans)
                y0 = min(s.y0 for s in matched_spans)
                x1 = max(s.x1 for s in matched_spans)
                y1 = max(s.y1 for s in matched_spans)
            else:
                x0 = y0 = x1 = y1 = 0.0

            confidence = nm.confidence
            status = (
                RedactionStatus.AUTO_REDACT
                if confidence >= auto_thresh
                else RedactionStatus.FLAGGED
            )

            candidates.append(RedactionCandidate(
                text=entity_text,
                category=PIICategory.PERSON_NAME,
                status=status,
                confidence=confidence,
                page_num=page_num,
                x0=x0, y0=y0, x1=x1, y1=y1,
                reason=nm.reason,
                source_file=source_filename,
            ))

        # ── Regex detection ───────────────────────────────────────────────
        regex_matches = find_regex_matches(page_text, page_num)

        for rm in regex_matches:
            entity_text = rm["text"]

            if is_subject_match(entity_text, subject):
                candidates.append(RedactionCandidate(
                    text=entity_text,
                    category=rm["category"],
                    status=RedactionStatus.EXCLUDED_SUBJECT,
                    confidence=1.0,
                    page_num=page_num,
                    reason=f"Matched data subject ({rm['category'].value})",
                    source_file=source_filename,
                ))
                continue

            matched_spans = map_text_to_spans(
                page_text, page_spans, rm["start"], rm["end"]
            )

            if matched_spans:
                x0 = min(s.x0 for s in matched_spans)
                y0 = min(s.y0 for s in matched_spans)
                x1 = max(s.x1 for s in matched_spans)
                y1 = max(s.y1 for s in matched_spans)
            else:
                x0 = y0 = x1 = y1 = 0.0

            confidence = rm["confidence"]
            status = (
                RedactionStatus.AUTO_REDACT
                if confidence >= auto_thresh
                else RedactionStatus.FLAGGED
            )

            candidates.append(RedactionCandidate(
                text=entity_text,
                category=rm["category"],
                status=status,
                confidence=confidence,
                page_num=page_num,
                x0=x0, y0=y0, x1=x1, y1=y1,
                reason=rm["reason"],
                source_file=source_filename,
            ))

        # ── Custom word detection ─────────────────────────────────────────────
        for cw in custom_words:
            phrase = cw["phrase"]
            flags = 0 if cw.get("case_sensitive", False) else re.IGNORECASE
            pattern = re.compile(r'\b' + re.escape(phrase) + r'\b', flags)
            for match in pattern.finditer(page_text):
                matched_spans = map_text_to_spans(
                    page_text, page_spans, match.start(), match.end()
                )
                if matched_spans:
                    x0 = min(s.x0 for s in matched_spans)
                    y0 = min(s.y0 for s in matched_spans)
                    x1 = max(s.x1 for s in matched_spans)
                    y1 = max(s.y1 for s in matched_spans)
                else:
                    x0 = y0 = x1 = y1 = 0.0

                candidates.append(RedactionCandidate(
                    text=match.group(0),
                    category=PIICategory.CUSTOM_WORD,
                    status=RedactionStatus.AUTO_REDACT,
                    confidence=1.0,
                    page_num=page_num,
                    x0=x0, y0=y0, x1=x1, y1=y1,
                    reason=f"Custom redaction word/phrase",
                    source_file=source_filename,
                ))

    # Deduplicate: same text + page + file + category
    # Also filter out disabled categories (except excluded_subject/staff which are always kept)
    seen = set()
    deduped = []
    for c in candidates:
        # Filter by enabled categories (always keep excluded items for audit trail)
        if (c.category.value not in enabled and
                c.status not in (RedactionStatus.EXCLUDED_SUBJECT, RedactionStatus.EXCLUDED_STAFF)):
            continue
        key = (normalize(c.text), c.page_num, c.source_file, c.category)
        if key not in seen:
            seen.add(key)
            # Compute risk flags for all candidates
            if not c.risk_flags:
                c.risk_flags = check_text_for_risk(c.text)
            deduped.append(c)

    return deduped

File: web/js/core/detector.js
// detector.js — Ported from sar/detector.py
// Main PII detection orchestrator.
// Attaches all exports to window.SARCore namespace.

window.SARCore = window.SARCore || {};

// ── Staff label patterns ──────────────────────────────────────────────────────
// Used to detect names that appear in staff-labelled positions (Practitioner:,
// Record author:, etc.) so they can be excluded as staff rather than redacted.

// Matches the label prefix (case-insensitive), used to look back before a name.
var _STAFF_LABEL_RE = /(?:Practitioner|Record\s+author|Filed\s+by|Requested\s+by|Authorised\s+by|Discontinued\s+by|Performed\s+by|Cancellation\s+reason):\s*(?:(?:Mr|Mrs|Ms|Miss|Dr|Prof|Sister|Nurse|Rev|Mx)\.?\s+)*/i;

// Name component: [A-Z][a-z']+(?:-[A-Z][a-z']*)? (handles Smith-Jones)
var _NAME_COMPONENT = "[A-Z][a-z']+(?:-[A-Z][a-z']*)?";

// Extracts the staff name that follows a staff label field.
// Stops at § (field separator), newline, "at " (org separator), or end of string.
var _STAFF_NAME_EXTRACT_RE = new RegExp(
    '(?:Practitioner|Record\\s+author|Filed\\s+by|Requested\\s+by|' +
    'Authorised\\s+by|Discontinued\\s+by|Performed\\s+by)\\s*:\\s*' +
    '((?:(?:Mr|Mrs|Ms|Miss|Dr|Prof|Sister|Nurse|Rev|Mx)\\.?\\s+)?' +
    '(?:' + _NAME_COMPONENT + ')(?:[ \\t]+(?:' + _NAME_COMPONENT + ')){0,3})' +
    '(?=[ \\t]*(?:\\xa7|\\n|$|at\\s+[A-Z]))',
    'gi'
);

// ── Internal helpers ──────────────────────────────────────────────────────────

function _normalize(text) {
    return text.trim().toLowerCase();
}

function _stripPunctuation(text) {
    return text.replace(/^[\s.,;:'"\\-]+|[\s.,;:'"\\-]+$/g, '');
}

/**
 * Extract names that appear in staff-labelled positions on this page.
 * e.g. "Practitioner: Dr David Triska" → Set {"david triska", "dr david triska"}
 * @param {string} pageText
 * @returns {Set<string>}
 */
function _extractPageStaffNames(pageText) {
    var staff = new Set();
    var titles = new Set(["mr","mrs","ms","miss","dr","prof","sister","nurse","rev","mx"]);
    var re = new RegExp(_STAFF_NAME_EXTRACT_RE.source, 'gi');
    var m;
    while ((m = re.exec(pageText)) !== null) {
        var raw = m[1].trim();
        staff.add(_normalize(raw));
        // Also add without title prefix
        var parts = raw.split(/\s+/);
        while (parts.length > 0 && titles.has(parts[0].toLowerCase().replace(/\.$/, ''))) {
            parts.shift();
        }
        if (parts.length > 0) {
            staff.add(_normalize(parts.join(' ')));
        }
    }
    return staff;
}

/**
 * Check if the name at nameStart is directly preceded by a staff label.
 * Looks back up to 90 chars.
 * @param {string} pageText
 * @param {number} nameStart
 * @returns {boolean}
 */
function _isInStaffLabelContext(pageText, nameStart) {
    var pre = pageText.slice(Math.max(0, nameStart - 90), nameStart);
    return _STAFF_LABEL_RE.test(pre);
}

// ── isSubjectMatch ────────────────────────────────────────────────────────────

/**
 * Check if detected text matches the data subject's known details.
 * @param {string} entityText
 * @param {object} subject  SubjectDetails object
 * @returns {boolean}
 */
function isSubjectMatch(entityText, subject) {
    var cleaned = _normalize(_stripPunctuation(entityText));

    // Full name check
    if (subject.full_name) {
        if (cleaned === _normalize(subject.full_name)) return true;
        if (_normalize(subject.full_name).length > 0 && cleaned.includes(_normalize(subject.full_name))) return true;
    }

    // First name (only match single-word entities)
    if (subject.first_name) {
        var entityWords = cleaned.split(/\s+/);
        if (entityWords.length === 1 && entityWords[0] === _normalize(subject.first_name)) {
            return true;
        }
    }

    // Last name: match lone surname OR full name where first word also matches subject
    if (subject.last_name) {
        var subjLast = _normalize(subject.last_name);
        var entityWords2 = cleaned.split(/\s+/);
        if (entityWords2.includes(subjLast)) {
            if (entityWords2.length === 1) return true; // Lone surname
            if (subject.first_name && entityWords2.includes(_normalize(subject.first_name))) {
                return true; // Both first & last match subject
            }
        }
    }

    // Aliases
    if (subject.aliases) {
        for (var i = 0; i < subject.aliases.length; i++) {
            var alias = subject.aliases[i];
            if (alias && cleaned === _normalize(alias)) return true;
        }
    }

    // NHS number
    if (subject.nhs_number) {
        var strippedEntity = entityText.replace(/[\s\-]/g, '');
        var strippedSubject = subject.nhs_number.replace(/[\s\-]/g, '');
        if (strippedEntity === strippedSubject) return true;
    }

    // Date of birth
    if (subject.date_of_birth && cleaned === _normalize(subject.date_of_birth)) {
        return true;
    }

    // Phone
    if (subject.phone) {
        var strippedEntityPhone = entityText.replace(/[\s\-()]/g, '');
        var strippedSubjectPhone = subject.phone.replace(/[\s\-()]/g, '');
        if (strippedEntityPhone === strippedSubjectPhone) return true;
    }

    // Email
    if (subject.email && cleaned === _normalize(subject.email)) {
        return true;
    }

    // Address (70% word overlap)
    if (subject.address) {
        var stopwords = new Set(["the","a","an","and","of","in","at","on","to"]);
        var addrWords = new Set(_normalize(subject.address).split(/\s+/));
        var entityWordsSet = new Set(cleaned.split(/\s+/).filter(function(w) {
            return !stopwords.has(w);
        }));
        if (entityWordsSet.size > 0) {
            var overlap = 0;
            entityWordsSet.forEach(function(w) {
                if (addrWords.has(w)) overlap++;
            });
            if (overlap >= Math.max(1, entityWordsSet.size * 0.7)) return true;
        }
    }

    return false;
}

// ── isStaffName ───────────────────────────────────────────────────────────────

/**
 * Check if a detected name matches any member of the staff list.
 * Uses normalized comparison on full name and individual components.
 * @param {string} detectedName
 * @param {Array} staffList  Array of {name, role} objects
 * @returns {boolean}
 */
function isStaffName(detectedName, staffList) {
    if (!staffList || staffList.length === 0) return false;
    var titles = new Set(["dr","mr","mrs","ms","miss","prof","professor","nurse","sister"]);
    var normDetected = detectedName.trim().toLowerCase();
    var detectedParts = new Set(normDetected.split(/\s+/));

    for (var i = 0; i < staffList.length; i++) {
        var member = staffList[i];
        var staffName = member.name.trim().toLowerCase();
        var staffParts = new Set(staffName.split(/\s+/));

        // Exact match
        if (normDetected === staffName) return true;

        // Detected name is a component of a staff name
        if (staffParts.has(normDetected)) return true;

        // Meaningful overlap (excluding titles)
        var meaningfulDetected = new Set();
        detectedParts.forEach(function(p) { if (!titles.has(p)) meaningfulDetected.add(p); });
        var meaningfulStaff = new Set();
        staffParts.forEach(function(p) { if (!titles.has(p)) meaningfulStaff.add(p); });

        if (meaningfulDetected.size > 0 && meaningfulStaff.size > 0) {
            var hasOverlap = false;
            meaningfulDetected.forEach(function(p) {
                if (meaningfulStaff.has(p)) hasOverlap = true;
            });
            if (hasOverlap) return true;
        }
    }
    return false;
}

// ── mapTextToSpans ────────────────────────────────────────────────────────────

/**
 * Find TextSpan objects whose bounding boxes correspond to the given
 * character range in pageText.
 * @param {string} pageText
 * @param {Array} pageSpans  Array of TextSpan objects
 * @param {number} charStart
 * @param {number} charEnd
 * @returns {Array}
 */
function mapTextToSpans(pageText, pageSpans, charStart, charEnd) {
    var matching = [];
    var searchFrom = 0;

    for (var i = 0; i < pageSpans.length; i++) {
        var span = pageSpans[i];
        var pos = pageText.indexOf(span.text, searchFrom);
        if (pos === -1) {
            pos = pageText.indexOf(span.text);
        }
        if (pos === -1) continue;

        var spanEnd = pos + span.text.length;
        searchFrom = pos + 1;

        if (pos < charEnd && spanEnd > charStart) {
            matching.push(span);
        }
    }
    return matching;
}

// ── checkTextForRisk ──────────────────────────────────────────────────────────

/**
 * Check text against risk word dictionary.
 * Returns array of {phrase, category} matches.
 * @param {string} text
 * @param {object} riskWords  {category: [words...]}
 * @returns {Array}
 */
function checkTextForRisk(text, riskWords) {
    if (!text || !riskWords) return [];
    var results = [];
    var categories = Object.keys(riskWords);
    for (var i = 0; i < categories.length; i++) {
        var category = categories[i];
        var words = riskWords[category];
        for (var j = 0; j < words.length; j++) {
            var word = words[j];
            var escaped = word.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
            var re = new RegExp('\\b' + escaped + '\\b', 'i');
            if (re.test(text)) {
                results.push({ phrase: word, category: category });
            }
        }
    }
    return results;
}

// ── detectPII ─────────────────────────────────────────────────────────────────

/**
 * Detect third-party PII in page texts.
 * Combines rule-based name detection with regex patterns.
 *
 * @param {Array} pageTexts      Array of strings, one per page (0-indexed)
 * @param {Array} textSpansByPage  Array of arrays of TextSpan objects, one per page
 * @param {object} subject       SubjectDetails object
 * @param {Array} staffList      Array of {name, role} staff objects
 * @param {Array} customWords    Array of {phrase, case_sensitive} objects
 * @param {object} settings      DetectionSettings object
 * @param {object} riskWords     {category: [words...]} risk word dictionary
 * @param {string} sourceFilename
 * @returns {Array}  Array of RedactionCandidate objects
 */
function detectPII(pageTexts, textSpansByPage, subject, staffList, customWords, settings, riskWords, sourceFilename) {
    var PIICategory    = window.SARCore.PIICategory;
    var RedactionStatus = window.SARCore.RedactionStatus;
    var createRedactionCandidate = window.SARCore.createRedactionCandidate;
    var detectNames    = window.SARCore.detectNames;
    var findRegexMatches = window.SARCore.findRegexMatches;

    if (!settings) {
        settings = window.SARCore.createDetectionSettings();
    }

    var autoThresh = settings.auto_redact_threshold;
    var flagThresh = settings.flag_threshold;
    var enabled = new Set(settings.enabled_categories);
    var candidates = [];
    var titles = new Set(["mr","mrs","ms","miss","dr","prof","sister","nurse","rev","mx"]);

    for (var pageNum = 0; pageNum < pageTexts.length; pageNum++) {
        var pageText = pageTexts[pageNum] || '';
        if (!pageText.trim()) continue;

        var pageSpans = (textSpansByPage && textSpansByPage[pageNum]) || [];

        // Extract practitioner/author names from labelled fields on this page
        var pageStaffNames = _extractPageStaffNames(pageText);

        // ── Name detection ─────────────────────────────────────────────────────
        var nameMatches = detectNames(pageText);

        for (var ni = 0; ni < nameMatches.length; ni++) {
            var nm = nameMatches[ni];
            var entityText = _stripPunctuation(nm.text);
            if (!entityText) continue;

            if (isSubjectMatch(entityText, subject)) {
                candidates.push(createRedactionCandidate({
                    text:        entityText,
                    category:    PIICategory.PERSON_NAME,
                    status:      RedactionStatus.EXCLUDED_SUBJECT,
                    confidence:  1.0,
                    page_num:    pageNum,
                    reason:      'Matched data subject',
                    source_file: sourceFilename || '',
                }));
                continue;
            }

            // Check staff list and contextual staff names
            var entityNorm = _normalize(entityText);
            var entityWords = entityNorm.split(/\s+/);
            var entityNoTitle = entityWords.filter(function(w, i) {
                return !(i === 0 && titles.has(w.replace(/\.$/, '')));
            }).join(' ');

            var inStaffLabel = _isInStaffLabelContext(pageText, nm.start);
            var inPageStaff = pageStaffNames.has(entityNorm) || pageStaffNames.has(entityNoTitle);

            if (isStaffName(entityText, staffList) || inStaffLabel || inPageStaff) {
                var staffReason = 'Matched staff list';
                if (inStaffLabel) {
                    staffReason = 'Name in staff-labelled field (Practitioner/Author)';
                } else if (inPageStaff) {
                    staffReason = 'Name matched practitioner on this page';
                }
                candidates.push(createRedactionCandidate({
                    text:        entityText,
                    category:    PIICategory.PERSON_NAME,
                    status:      RedactionStatus.EXCLUDED_STAFF,
                    confidence:  1.0,
                    page_num:    pageNum,
                    reason:      staffReason,
                    source_file: sourceFilename || '',
                }));
                continue;
            }

            var matchedSpans = mapTextToSpans(pageText, pageSpans, nm.start, nm.end);
            var x0, y0, x1, y1;
            if (matchedSpans.length > 0) {
                x0 = Math.min.apply(null, matchedSpans.map(function(s) { return s.x0; }));
                y0 = Math.min.apply(null, matchedSpans.map(function(s) { return s.y0; }));
                x1 = Math.max.apply(null, matchedSpans.map(function(s) { return s.x1; }));
                y1 = Math.max.apply(null, matchedSpans.map(function(s) { return s.y1; }));
            } else {
                x0 = y0 = x1 = y1 = 0;
            }

            var confidence = nm.confidence;
            var status = (confidence >= autoThresh)
                ? RedactionStatus.AUTO_REDACT
                : RedactionStatus.FLAGGED;

            candidates.push(createRedactionCandidate({
                text:        entityText,
                category:    PIICategory.PERSON_NAME,
                status:      status,
                confidence:  confidence,
                page_num:    pageNum,
                x0: x0, y0: y0, x1: x1, y1: y1,
                reason:      nm.reason,
                source_file: sourceFilename || '',
            }));
        }

        // ── Regex detection ────────────────────────────────────────────────────
        var regexMatches = findRegexMatches(pageText, pageNum);

        for (var ri = 0; ri < regexMatches.length; ri++) {
            var rm = regexMatches[ri];
            var rEntityText = rm.text;

            if (isSubjectMatch(rEntityText, subject)) {
                candidates.push(createRedactionCandidate({
                    text:        rEntityText,
                    category:    rm.category,
                    status:      RedactionStatus.EXCLUDED_SUBJECT,
                    confidence:  1.0,
                    page_num:    pageNum,
                    reason:      'Matched data subject (' + rm.category + ')',
                    source_file: sourceFilename || '',
                }));
                continue;
            }

            var rMatchedSpans = mapTextToSpans(pageText, pageSpans, rm.start, rm.end);
            var rx0, ry0, rx1, ry1;
            if (rMatchedSpans.length > 0) {
                rx0 = Math.min.apply(null, rMatchedSpans.map(function(s) { return s.x0; }));
                ry0 = Math.min.apply(null, rMatchedSpans.map(function(s) { return s.y0; }));
                rx1 = Math.max.apply(null, rMatchedSpans.map(function(s) { return s.x1; }));
                ry1 = Math.max.apply(null, rMatchedSpans.map(function(s) { return s.y1; }));
            } else {
                rx0 = ry0 = rx1 = ry1 = 0;
            }

            var rConf = rm.confidence;
            var rStatus = (rConf >= autoThresh)
                ? RedactionStatus.AUTO_REDACT
                : RedactionStatus.FLAGGED;

            candidates.push(createRedactionCandidate({
                text:        rEntityText,
                category:    rm.category,
                status:      rStatus,
                confidence:  rConf,
                page_num:    pageNum,
                x0: rx0, y0: ry0, x1: rx1, y1: ry1,
                reason:      rm.reason,
                source_file: sourceFilename || '',
            }));
        }

        // ── Custom word detection ──────────────────────────────────────────────
        if (customWords && customWords.length > 0) {
            for (var ci = 0; ci < customWords.length; ci++) {
                var cw = customWords[ci];
                var phrase = cw.phrase;
                var escaped = phrase.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
                var cwFlags = cw.case_sensitive ? 'g' : 'gi';
                var cwRe = new RegExp('\\b' + escaped + '\\b', cwFlags);
                var cwM;
                while ((cwM = cwRe.exec(pageText)) !== null) {
                    var cwSpans = mapTextToSpans(pageText, pageSpans, cwM.index, cwM.index + cwM[0].length);
                    var cwx0, cwy0, cwx1, cwy1;
                    if (cwSpans.length > 0) {
                        cwx0 = Math.min.apply(null, cwSpans.map(function(s) { return s.x0; }));
                        cwy0 = Math.min.apply(null, cwSpans.map(function(s) { return s.y0; }));
                        cwx1 = Math.max.apply(null, cwSpans.map(function(s) { return s.x1; }));
                        cwy1 = Math.max.apply(null, cwSpans.map(function(s) { return s.y1; }));
                    } else {
                        cwx0 = cwy0 = cwx1 = cwy1 = 0;
                    }
                    candidates.push(createRedactionCandidate({
                        text:        cwM[0],
                        category:    PIICategory.CUSTOM_WORD,
                        status:      RedactionStatus.AUTO_REDACT,
                        confidence:  1.0,
                        page_num:    pageNum,
                        x0: cwx0, y0: cwy0, x1: cwx1, y1: cwy1,
                        reason:      'Custom redaction word/phrase',
                        source_file: sourceFilename || '',
                    }));
                }
            }
        }
    }

    // ── Deduplicate & filter ───────────────────────────────────────────────────
    var seen = new Set();
    var deduped = [];
    var excludedStatuses = new Set([
        RedactionStatus.EXCLUDED_SUBJECT,
        RedactionStatus.EXCLUDED_STAFF,
    ]);

    for (var di = 0; di < candidates.length; di++) {
        var c = candidates[di];

        // Filter by enabled categories (always keep excluded items for audit trail)
        if (!enabled.has(c.category) && !excludedStatuses.has(c.status)) {
            continue;
        }

        var key = _normalize(c.text) + '|' + c.page_num + '|' + (c.source_file || '') + '|' + c.category;
        if (!seen.has(key)) {
            seen.add(key);
            // Compute risk flags
            if (!c.risk_flags || c.risk_flags.length === 0) {
                c.risk_flags = checkTextForRisk(c.text, riskWords);
            }
            deduped.push(c);
        }
    }

    return deduped;
}


// ── Exports ───────────────────────────────────────────────────────────────────

window.SARCore.isSubjectMatch    = isSubjectMatch;
window.SARCore.isStaffName       = isStaffName;
window.SARCore.mapTextToSpans    = mapTextToSpans;
window.SARCore.checkTextForRisk  = checkTextForRisk;
window.SARCore.detectPII         = detectPII;
